File: ocaml/chap02/stack.ml
exception EMPTY

type 'a t = Nil | Cons of 'a * 'a t

let empty = Nil
let isEmpty = function
    Nil -> true
  | _ -> false

let cons (x, s) = Cons (x, s)
let head = function
    Nil -> raise EMPTY
  | Cons (x, _) -> x
let tail = function
    Nil -> raise EMPTY
  | Cons (_, s) -> s

File: rust/src/chap02/stack.rs
use std::rc::Rc;

// 図2.1
pub trait Stack<T>: Sized {
    fn empty() -> Self;
    fn is_empty(&self) -> bool;

    fn cons(&self, x: T) -> Self;
    fn head(&self) -> Rc<T>;
    fn tail(&self) -> Self;
}
